### DISCUSSION TITLE
custom parser does not work

### DISCUSSION THREAD

**Discussion by sleepingbel** [Security-Onion-Solutions/securityonion#2323] [Unsupported Versions] [answered]
Hello,

ISO install SO 2.3.10 Standalone
8 vCore
16 vRAM
2 TB Disk space

In SO 16.04 the syslog logs of devices like a router, printer and switch where parsed and marked as a log node, in SO 2.3.x that is no longer the case.

Therefore I made a custom syslog parser for standard syslogs of devices like a router, printer and a homelab esxi.

I have followed the following documentations:
- [https://docs.securityonion.net/en/2.3/elasticsearch.html#parsing](url) 
- [https://www.elastic.co/guide/en/elasticsearch/reference/master/ingest.html](url)

I used the syslog parser on location /opt/so/conf/elasticsearch/ingest/syslog as an example for understanding wat is hapening during the parsing.
For error checking the parser I used the Dev Tool "Grok Debugger" and in ingest node pipelines "test pipeline" .
On restarting elasticsearch there where no errors.

but when I look at the syslogs in kibana it remains as before and these logs are only parsed by default.
Can someone tell me what I'm doing wrong? Do I have to create a new index_setting for this? Because that is not mentioned anywhere in the SO or elastic documentation.

I see the parser in Ingest node pipelines
![afbeelding](https://user-images.githubusercontent.com/29328171/101986740-1d550380-3c90-11eb-94dc-4f37596c72bf.png)

This is the parser script
{
  "description" : "syslog",  "processors" : [
  {
    "dissect": {      "field": "message",      "pattern": "%{message}",      "on_failure": [        {          "drop": {}        }      ]    }
  },
  {
    "remove": {      "field": [        "type",        "agent"      ],      "ignore_failure": true    }
  },
  {
    "grok": {      "field": "message",      "patterns": [
        "^<%{INT:syslog.priority}>%{SYSLOGTIMESTAMP:syslog.timestamp} %{SYSLOGHOST:syslog.facility} %{SYSLOGPROG:syslog.program1} %{GREEDYDATA:real_message}$",
        "^<%{POSINT:priority}> %{SYSLOGLINE}$",
        "^<%{POSINT:priority}>%{SYSLOGLINE}$"
                                                        ],      "ignore_failure": true    }
  },
  {
    "remove": {      "field": [        "syslog.program1"      ],      "ignore_failure": true    }
  },
  {
    "rename": {      "field": "timestamp8601",      "target_field": "syslog.timestamp",      "ignore_missing": true    }
  },
  {
    "rename": {      "field": "timestamp",      "target_field": "syslog.timestamp",      "ignore_missing": true    }
  },
  {
    "rename": {      "field": "logsource",      "target_field": "syslog.facility",      "ignore_missing": true    }
  },
  {
    "rename": {      "field": "priority",      "target_field": "syslog.priority",      "ignore_missing": true    }
  },
  {
    "rename": {      "field": "program",      "target_field": "syslog.program",      "ignore_missing": true    }
  },
  {
    "remove": {      "field": [        "program"      ],      "ignore_failure": true    }
  },
  {
    "set": {      "field": "ingest.timestamp",      "value": "{{ @timestamp }}"    }
  },
  { 
    "pipeline":       { "name": "common" , "ignore_failure": true                                        } 
  }
                                                ]
}

These are the syslog inputs
![esxi](https://user-images.githubusercontent.com/29328171/101986620-327d6280-3c8f-11eb-9b1b-b1cf93106ceb.PNG)
![router](https://user-images.githubusercontent.com/29328171/101986623-37421680-3c8f-11eb-9329-87cc2216dbb6.PNG)
![printer](https://user-images.githubusercontent.com/29328171/101986627-3ad59d80-3c8f-11eb-92dd-dec3bbd826cd.PNG)

Thx in advance for all the help. 
In my spare time I am doing an evaluation of SO at home, this to make a complete report for my management in the hoop that they will implement SO after reading the report.

Bart

**Comment by sleepingbel** [comment]
I have found the solution, you must give your custom script the same name as the original script. In my case the custum script name must be syslog. My syslog dashboard is starting to fill up.

![Dashboard](https://user-images.githubusercontent.com/29328171/102015796-f6650300-3d5d-11eb-9f71-77efa3753091.PNG)

Maybe this can be mentioned in the documentation?

But still the IP address is not displayed correctly, I have seen in the Index Patterns that the field log.source.address is a text field.

![IP](https://user-images.githubusercontent.com/29328171/102015898-75f2d200-3d5e-11eb-81f1-9a0984c7187c.PNG)

Can I, in the same custum parser for the message, change the IP-adress? 

Also for the dashboard the Destination Ports sub-dashboard can be changed to Source Ports, because in SO you will receive all syslogs on port 514.

if you send all logs without having to parse them on the device, you have an advantage in that you only have to learn 1 parser language (elastic).

**Comment by dougburks** [comment, marked as answer]
You may want to change your parser so that it parses out source ip, source port, destination ip, and destination port and renames those fields to match Elastic Common Schema:
source.ip
source.port
destination.ip
destination.port

If you'd like some examples you can compare to, take a look at our existing Zeek parsers:
https://docs.securityonion.net/en/2.3/zeek-fields.html
